server/tests/common/replicated_loglet.rs

```diff
@@ -0,0 +1,170 @@
+    config.common.force_node_id = Some(node_id);
+    config.common.set_node_name(node_name);
+    config.common.set_base_dir(cluster.base_dir());
+    config
+        .common
+        .set_cluster_name(cluster.cluster_name().to_owned());
+    config.common.advertised_address = AdvertisedAddress::Uds(node_socket.clone());
+    config.common.bind_address = BindAddress::Uds(node_socket.clone());
+    config.common.metadata_store_client = cluster.nodes[0]
+        .config()
+        .common
+        .metadata_store_client
+        .clone();
+
+    restate_types::config::set_current_config(config.clone());
+
+    let node = restate_node::Node::create(Live::from_value(config)).await?;
+
+    let bifrost = node.bifrost();
+    let metadata_writer = node.metadata_writer();
+    let metadata_store_client = node.metadata_store_client();
+
+    node.start().await.into_test_result()?;
+
+    let loglet = bifrost
+        .find_tail_loglet(LogId::MIN, FindTailAttributes::default())
+        .await?;
+
+    Ok((bifrost, loglet, metadata_writer, metadata_store_client))
+}
+
+pub struct TestEnv {
+    bifrost: Bifrost,
+    pub loglet: Arc<dyn Loglet>,
+    metadata_writer: MetadataWriter,
+    metadata_store_client: MetadataStoreClient,
+    pub cluster: StartedCluster,
+    pub loglet_params: String,
+}
+
+impl TestEnv {
+    pub fn bifrost_admin(&self) -> BifrostAdmin<'_> {
+        BifrostAdmin::new(
+            &self.bifrost,
+            &self.metadata_writer,
+            &self.metadata_store_client,
+        )
+    }
+}
+
+pub async fn run_in_test_env<F, O>(
+    mut base_config: Configuration,
+    sequencer: GenerationalNodeId,
+    replication: ReplicationProperty,
+    log_server_count: u32,
+    mut future: F,
+) -> googletest::Result<()>
+where
+    F: FnMut(TestEnv) -> O,
+    O: std::future::Future<Output = googletest::Result<()>> + Send,
+{
+    let loglet_params = ReplicatedLogletParams {
+        loglet_id: ReplicatedLogletId::new(1),
+        sequencer,
+        replication,
+        nodeset: (1..log_server_count).collect(),
+        write_set: None,
+    };
+    let loglet_params = loglet_params.serialize()?;
+    base_config.bifrost.default_provider = ProviderKind::Replicated;
+    base_config.bifrost.default_provider_config = Some(loglet_params.clone());
+    base_config
+        .common
+        .set_bootstrap_num_partitions(NonZeroU16::new(1).unwrap());
+
+    let nodes = Node::new_test_nodes_with_metadata(
+        base_config,
+        BinarySource::CargoTest,
+        enum_set!(Role::LogServer),
+        log_server_count,
+    );
+
+    let tc = TaskCenterBuilder::default()
+        .default_runtime_handle(tokio::runtime::Handle::current())
+        .ingress_runtime_handle(tokio::runtime::Handle::current())
+        .build()
+        .expect("task_center builds");
+
+    // ensure base dir lives longer than the node, otherwise it sees shutdown errors
+    // this will still respect LOCAL_CLUSTER_RUNNER_RETAIN_TEMPDIR=true
```

Comment:
Q: Not sure if the cluster utility already does that or not, but should we print on stdout this base_dir path after the test is done (or before)?

Reply:
it does indeed already do it